File: architecture/cqrs/go/projector.go
package main

import (
	"context"

	"github.com/jackc/pgx/v5/pgxpool"
)

// setupReadModel создаёт read-сторону: денормализованную проекцию orders_read
// (заточена под запрос «заказы пользователя» — индекс по user_id) и таблицу чекпоинта
// projection_checkpoint (позиция, до которой проектор уже применил лог).
//
// Ключевой инвариант: чекпоинт двигается в ОДНОЙ транзакции с апдейтом проекции
// (см. projectOnce). Иначе при сбое между «применил» и «записал позицию» события
// либо потеряются, либо применятся дважды.
func setupReadModel(ctx context.Context, pool *pgxpool.Pool) error {
	_, err := pool.Exec(ctx, `
		DROP TABLE IF EXISTS orders_read;
		CREATE TABLE orders_read (
			order_id    BIGINT PRIMARY KEY,
			user_id     TEXT   NOT NULL,
			status      TEXT   NOT NULL,
			amount      BIGINT NOT NULL,
			updated_seq BIGINT NOT NULL
		);
		CREATE INDEX ON orders_read (user_id);   -- проекция под конкретный запрос: список заказов юзера

		DROP TABLE IF EXISTS projection_checkpoint;
		CREATE TABLE projection_checkpoint (
			name     TEXT PRIMARY KEY,
			last_seq BIGINT NOT NULL
		);
		INSERT INTO projection_checkpoint (name, last_seq) VALUES ('orders_read', 0);
	`)
	return err
}

// projectOnce применяет к проекции все события лога с seq > чекпоинта и атомарно
// сдвигает чекпоинт. Применение идемпотентно (UPSERT по order_id), поэтому повтор
// того же батча после сбоя не ломает проекцию. Возвращает число применённых событий.
//
// Именно то, что этот шаг вызывается ОТДЕЛЬНО от команд записи, и создаёт лаг:
// пока projectOnce не позвали, свежие события уже в логе, но проекция их не видит.
func projectOnce(ctx context.Context, pool *pgxpool.Pool) (applied int, err error) {
	return applyLog(ctx, pool, "orders_read", "projection_checkpoint")
}

// applyLog — общая механика «применить хвост лога в таблицу-проекцию + сдвинуть
// чекпоинт в той же транзакции». Используется и штатным проектором, и rebuild-ом
// (там применяет в orders_read_v2 со своим чекпоинтом).
func applyLog(ctx context.Context, pool *pgxpool.Pool, table, ckpt string) (applied int, err error) {
	tx, err := pool.Begin(ctx)
	if err != nil {
		return 0, err
	}
	defer tx.Rollback(ctx)

	var from int64
	if err = tx.QueryRow(ctx,
		`SELECT last_seq FROM `+ckpt+` WHERE name=$1 FOR UPDATE`, table).Scan(&from); err != nil {
		return 0, err
	}

	rows, err := tx.Query(ctx,
		`SELECT seq, order_id, user_id, type, amount FROM order_events
		  WHERE seq > $1 ORDER BY seq`, from)
	if err != nil {
		return 0, err
	}
	type ev struct {
		seq, orderID, amount int64
		userID, typ          string
	}
	var evs []ev
	for rows.Next() {
		var e ev
		if err = rows.Scan(&e.seq, &e.orderID, &e.userID, &e.typ, &e.amount); err != nil {
			rows.Close()
			return 0, err
		}
		evs = append(evs, e)
	}
	rows.Close()
	if err = rows.Err(); err != nil {
		return 0, err
	}

	var last int64 = from
	for _, e := range evs {
		switch e.typ {
		case "created":
			// Идемпотентный UPSERT: повторное применение того же события безопасно.
			_, err = tx.Exec(ctx,
				`INSERT INTO `+table+` (order_id, user_id, status, amount, updated_seq)
				 VALUES ($1, $2, 'new', $3, $4)
				 ON CONFLICT (order_id) DO UPDATE
				   SET status=EXCLUDED.status, amount=EXCLUDED.amount, updated_seq=EXCLUDED.updated_seq`,
				e.orderID, e.userID, e.amount, e.seq)
		case "paid":
			_, err = tx.Exec(ctx,
				`UPDATE `+table+` SET status='paid', updated_seq=$2 WHERE order_id=$1`,
				e.orderID, e.seq)
		}
		if err != nil {
			return 0, err
		}
		last = e.seq
	}

	if _, err = tx.Exec(ctx,
		`UPDATE `+ckpt+` SET last_seq=$2 WHERE name=$1`, table, last); err != nil {
		return 0, err
	}
	if err = tx.Commit(ctx); err != nil {
		return 0, err
	}
	return len(evs), nil
}

// checkpointSeq — до какого seq проектор уже применил лог.
func checkpointSeq(ctx context.Context, pool *pgxpool.Pool) (int64, error) {
	var s int64
	err := pool.QueryRow(ctx,
		`SELECT last_seq FROM projection_checkpoint WHERE name='orders_read'`).Scan(&s)
	return s, err
}

// projectionLag — метрика отставания проекции: хвост лога минус чекпоинт проектора.
// 0 — проекция догнала лог; >0 — есть непроapplication-ные события (окно, в котором
// ломается read-your-writes).
func projectionLag(ctx context.Context, pool *pgxpool.Pool) (int64, error) {
	tail, err := tailSeq(ctx, pool)
	if err != nil {
		return 0, err
	}
	ck, err := checkpointSeq(ctx, pool)
	if err != nil {
		return 0, err
	}
	return tail - ck, nil
}

// readUserOrdersFromProjection — read-путь через проекцию (быстро, денормализовано,
// но может отставать). Для «чужих» заказов этого достаточно; для собственных свежих —
// нет (см. read-your-writes в main.go).
func readUserOrdersFromProjection(ctx context.Context, pool *pgxpool.Pool, userID string) ([]Order, error) {
	rows, err := pool.Query(ctx,
		`SELECT order_id, user_id, amount, status, updated_seq
		   FROM orders_read WHERE user_id=$1 ORDER BY order_id`, userID)
	if err != nil {
		return nil, err
	}
	return scanOrders(rows)
}

// waitForProjection — второй приём read-your-writes: «подождать по токену». Крутит
// проектор, пока чекпоинт не дойдёт до нужного seq, после чего из проекции уже видно
// собственную запись. В проде это ожидание готовности проекции (а не busy-loop).
func waitForProjection(ctx context.Context, pool *pgxpool.Pool, token int64) error {
	for {
		ck, err := checkpointSeq(ctx, pool)
		if err != nil {
			return err
		}
		if ck >= token {
			return nil
		}
		if _, err := projectOnce(ctx, pool); err != nil {
			return err
		}
	}
}

File: architecture/cqrs/java/src/main/java/tech/khorost/cqrs/Projector.java
package tech.khorost.cqrs;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

/**
 * Read-сторона CQRS: async-проекция {@code orders_read} (денормализована и заточена под
 * запрос «список заказов пользователя» — индекс по user_id) плюс таблица чекпоинта
 * {@code projection_checkpoint} (позиция, до которой проектор уже применил лог).
 * <p>
 * Ключевой инвариант: чекпоинт двигается в ОДНОЙ транзакции с апдейтом проекции (см.
 * {@link #applyLog}). Иначе при сбое между «применил» и «записал позицию» события либо
 * потеряются, либо применятся дважды.
 */
final class Projector {

    private final Connection db;

    Projector(Connection db) {
        this.db = db;
    }

    /** Создаёт read-сторону: проекцию orders_read (индекс по user_id) + чекпоинт позиции. */
    void setup() throws SQLException {
        try (Statement st = db.createStatement()) {
            st.execute("""
                DROP TABLE IF EXISTS orders_read;
                CREATE TABLE orders_read (
                    order_id    BIGINT PRIMARY KEY,
                    user_id     TEXT   NOT NULL,
                    status      TEXT   NOT NULL,
                    amount      BIGINT NOT NULL,
                    updated_seq BIGINT NOT NULL
                );
                CREATE INDEX ON orders_read (user_id);   -- проекция под конкретный запрос: список заказов юзера

                DROP TABLE IF EXISTS projection_checkpoint;
                CREATE TABLE projection_checkpoint (
                    name     TEXT PRIMARY KEY,
                    last_seq BIGINT NOT NULL
                );
                INSERT INTO projection_checkpoint (name, last_seq) VALUES ('orders_read', 0);
                """);
        }
    }

    /**
     * Применяет к проекции все события лога с seq &gt; чекпоинта и атомарно сдвигает чекпоинт.
     * Применение идемпотентно (UPSERT по order_id), повтор того же батча после сбоя безопасен.
     * Именно то, что этот шаг вызывается ОТДЕЛЬНО от команд записи, и создаёт лаг.
     * Возвращает число применённых событий.
     */
    int projectOnce() throws SQLException {
        return applyLog("orders_read", "projection_checkpoint");
    }

    /**
     * Общая механика «применить хвост лога в таблицу-проекцию + сдвинуть чекпоинт в той же
     * транзакции». Используется и штатным проектором, и rebuild-ом (там применяет в
     * orders_read_v2 со своим чекпоинтом). Имена table/ckpt подставляются в SQL — это
     * доверенные внутренние константы, не пользовательский ввод.
     */
    int applyLog(String table, String ckpt) throws SQLException {
        db.setAutoCommit(false);
        try {
            long from;
            try (PreparedStatement ps = db.prepareStatement(
                    "SELECT last_seq FROM " + ckpt + " WHERE name=? FOR UPDATE")) {
                ps.setString(1, table);
                try (ResultSet rs = ps.executeQuery()) {
                    rs.next();
                    from = rs.getLong(1);
                }
            }

            record Ev(long seq, long orderId, String userId, String typ, long amount) {
            }
            List<Ev> evs = new ArrayList<>();
            try (PreparedStatement ps = db.prepareStatement(
                    "SELECT seq, order_id, user_id, type, amount FROM order_events "
                            + "WHERE seq > ? ORDER BY seq")) {
                ps.setLong(1, from);
                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) {
                        evs.add(new Ev(rs.getLong(1), rs.getLong(2), rs.getString(3),
                                rs.getString(4), rs.getLong(5)));
                    }
                }
            }

            long last = from;
            for (Ev e : evs) {
                switch (e.typ()) {
                    case "created" -> {
                        // Идемпотентный UPSERT: повторное применение того же события безопасно.
                        try (PreparedStatement ps = db.prepareStatement(
                                "INSERT INTO " + table + " (order_id, user_id, status, amount, updated_seq) "
                                        + "VALUES (?, ?, 'new', ?, ?) "
                                        + "ON CONFLICT (order_id) DO UPDATE "
                                        + "  SET status=EXCLUDED.status, amount=EXCLUDED.amount, "
                                        + "      updated_seq=EXCLUDED.updated_seq")) {
                            ps.setLong(1, e.orderId());
                            ps.setString(2, e.userId());
                            ps.setLong(3, e.amount());
                            ps.setLong(4, e.seq());
                            ps.executeUpdate();
                        }
                    }
                    case "paid" -> {
                        try (PreparedStatement ps = db.prepareStatement(
                                "UPDATE " + table + " SET status='paid', updated_seq=? WHERE order_id=?")) {
                            ps.setLong(1, e.seq());
                            ps.setLong(2, e.orderId());
                            ps.executeUpdate();
                        }
                    }
                    default -> {
                    }
                }
                last = e.seq();
            }

            try (PreparedStatement ps = db.prepareStatement(
                    "UPDATE " + ckpt + " SET last_seq=? WHERE name=?")) {
                ps.setLong(1, last);
                ps.setString(2, table);
                ps.executeUpdate();
            }
            db.commit();
            return evs.size();
        } catch (SQLException e) {
            db.rollback();
            throw e;
        } finally {
            db.setAutoCommit(true);
        }
    }

    /** До какого seq проектор уже применил лог. */
    long checkpointSeq() throws SQLException {
        try (Statement st = db.createStatement();
             ResultSet rs = st.executeQuery(
                     "SELECT last_seq FROM projection_checkpoint WHERE name='orders_read'")) {
            rs.next();
            return rs.getLong(1);
        }
    }

    /**
     * Метрика отставания проекции: хвост лога минус чекпоинт проектора. 0 — проекция догнала
     * лог; &gt;0 — есть непроигранные события (окно, в котором ломается read-your-writes).
     */
    long projectionLag(WriteModel write) throws SQLException {
        return write.tailSeq() - checkpointSeq();
    }

    /**
     * Read-путь через проекцию (быстро, денормализовано, но может отставать). Для «чужих»
     * заказов этого достаточно; для собственных свежих — нет (см. read-your-writes в Main).
     */
    List<Order> readUserOrdersFromProjection(String userId) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement(
                "SELECT order_id, user_id, amount, status, updated_seq "
                        + "FROM orders_read WHERE user_id=? ORDER BY order_id")) {
            ps.setString(1, userId);
            try (ResultSet rs = ps.executeQuery()) {
                return WriteModel.scanOrders(rs);
            }
        }
    }

    /**
     * Второй приём read-your-writes: «подождать по токену». Крутит проектор, пока чекпоинт не
     * дойдёт до нужного seq, после чего из проекции уже видно собственную запись. В проде это
     * ожидание готовности проекции (а не busy-loop).
     */
    void waitForProjection(long token) throws SQLException {
        while (checkpointSeq() < token) {
            projectOnce();
        }
    }
}
